corehq/apps/userreports/tests/test_json_query_expressions.py

```diff
@@ -0,0 +1,455 @@
+    ),
+    # if items_expression returns empty, an empty list is returned
+    (
+        {'items': [{'key': 'v1'}, {'no_key': 'v2'}]},
+        {'type': 'property_name', 'property_name': 'no_match'},
+        {'type': 'property_name', 'property_name': 'key'},
+        []
+    ),
+])
+def test_map_items_basic(self, doc, items_ex, map_ex, expected):
+    expression = ExpressionFactory.from_spec({
+        'type': 'map_items',
+        'items_expression': items_ex,
+        'map_expression': map_ex
+    })
+    self.assertEqual(expression(doc), expected)
+
+
+@generate_cases([
+    ([{'key': 'v1'}, {'key': 'v2'}], None),
+    ({'type': 'identity'}, {}),
+    ({'type': 'identity'}, 'invalid_agg_fn'),
+])
+def test_reduce_items_bad_spec(self, items_ex, reduce_ex):
+    expression = {
+        'type': 'reduce_items',
+    }
+    if items_ex is not None:
+        expression['items_expression'] = items_ex
+    if reduce_ex is not None:
+        expression['aggregation_fn'] = reduce_ex
+
+    with self.assertRaises(BadSpecError):
+        ExpressionFactory.from_spec(expression)(items_ex, reduce_ex)
+
+
+@generate_cases([
+    (
+        ['a', 'b'],
+        {'type': 'identity'},
+        'count',
+        2,
+    ),
+    (
+        {'items': ['a', 'b']},
+        {'type': 'property_name', 'property_name': 'items'},
+        'count',
+        2,
+    ),
+    (
+        {'items': ['a', 'b']},
+        {'type': 'property_name', 'property_name': 'no_match'},
+        'count',
+        0,
+    ),
+    (
+        [],
+        {'type': 'identity'},
+        'count',
+        0,
+    ),
+    (
+        [1, 2],
+        {'type': 'identity'},
+        'sum',
+        3
+    ),
+    (
+        [1, 2],
+        {'type': 'identity'},
+        'first_item',
+        1
+    ),
+    (
+        [1, 2],
+        {'type': 'identity'},
+        'last_item',
+        2
+    ),
+])
+def test_reduce_items_basic(self, doc, items_ex, reduce_ex, expected):
+    expression = ExpressionFactory.from_spec({
+        'type': 'reduce_items',
+        'items_expression': items_ex,
+        'aggregation_fn': reduce_ex
+    })
+    self.assertEqual(expression(doc), expected)
+
+
+@generate_cases([
+    ({}, ), (None, )
+])
+def test_flatten_bad_spec(self, items_ex):
+    with self.assertRaises(BadSpecError):
+        ExpressionFactory.from_spec({
+            'type': 'flatten',
+            'items_expression': items_ex
+        })
+
+
+@generate_cases([
+    (
+        {'k': [[1, 2], [4, 5]]},
+        {'type': 'property_name', 'property_name': 'k'},
+        [1, 2, 4, 5]
+    ),
+    (
+        {},
+        [[1, 2], [4, 5]],
+        [1, 2, 4, 5]
+    ),
+    (
+        [[1, 2], [4, 5]],
+        {'type': 'identity'},
+        [1, 2, 4, 5]
+    ),
+    (
+        [[1, 2], 4, 5],
+        {'type': 'identity'},
+        []
+    ),
+    (
+        ["a", "b"],
+        {'type': 'identity'},
+        ["a", "b"],
+    ),
+    (
+        [1, 2],
+        {'type': 'identity'},
+        [],
```

Comment:
the difference between these two is pretty subtle and weird. It's because strings are iterables of charactres whereas numbers aren't. was your intention that this was the correct way for it to behave or were you just documenting the oddity? It might make sense to somehow treat strings as non-lists and fail in the same way we'd fail with numbers.